dfs/download.go

```diff
@@ -30,6 +31,74 @@ var (
 // Download will download the image from upstream and save the image to the
 // registry.
 func (dfs *DistributedFilesystem) Download(image, tenantID string, upgrade bool) (string, error) {
+	if !upgrade {
+		// Is this an image that has already been deployed?
+		if rImage, err := dfs.findImage(image, tenantID); err != nil {
+			return "", err
+		} else if rImage != "" {
+			return rImage, nil
+		}
+	}
+	// Figure out what the image will be in the docker registry
+	rImage, err := dfs.parseRegistryImage(image, tenantID)
+	if err != nil {
+		return "", err
+	}
+	// Pull the image (if it doesn't exist locally)
+	img, err := dfs.pullImage(image)
+	if err != nil {
+		return "", err
+	}
+	rimg, err := dfs.index.FindImage(rImage)
+	if err == index.ErrImageNotFound {
+		// Image does not exist in the registry, so push
+		if err := dfs.index.PushImage(rImage, img.ID); err != nil {
+			glog.Errorf("Could not push image %s into registry: %s", rImage, err)
+			return "", err
+		}
+		return rImage, nil
+	} else if err != nil {
+		glog.Errorf("Could not look up image %s from the registry: %s", rImage, err)
+		return "", err
+	}
+	// Compare the uuids of the topmost layer
+	if rimg.UUID != img.ID {
+		if upgrade {
+			// We are upgrading the image, so overwrite the existing tag with
+			// the new UUID.
+			if err := dfs.index.PushImage(rImage, img.ID); err != nil {
+				glog.Errorf("Could not upgrade image %s into registry: %s", rImage, err)
+				return "", err
+			}
+		} else {
+			return "", ErrImageCollision
+		}
+	}
+	return rImage, nil
+}
+
+// findImage will verify whether the image has already been deployed with the
+// application.
+func (dfs *DistributedFilesystem) findImage(image, tenantID string) (string, error) {
+	// Is the image in the index?
+	rImage, err := dfs.index.FindImage(image)
+	if err != nil {
+		if err == index.ErrImageNotFound {
+			return "", nil
```

Comment:
Are we using a return value set of `"", nil` to indicate that an image was not found? Any reason not to use index.ErrImageNotFound to convey that?